### PR TITLE
Remove API calls from `definitions` package

### PR DESCRIPTION
**Pull request by colinking**
## Description

This PR removes all API client usage from the `definitions` package. This was only used for fetching resources, which we now pass in directly. This lets the caller decide if an API call is necessary.

These changes are useful in studio, where we perform operations on definitions from within the CLI's API.

## Test plan

Tests should pass.

### CODE REVIEW

**Comment by colinking** [comment]
Current dependencies on/for this PR:
* main
  * **PR #567** <a href="https://app.graphite.dev/github/pr/airplanedev/lib/567" target="_blank">Graphite</a> 
    * **PR #568** <a href="https://app.graphite.dev/github/pr/airplanedev/lib/568" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/airplanedev/lib/568?utm_source=stack-comment).

File: pkg/deploy/taskdir/definitions/task_def.go [hunk 1]
```diff
@@ -52,8 +51,8 @@ type Definition struct {
 }
 
 type taskKind interface {
-	fillInUpdateTaskRequest(context.Context, api.IAPIClient, *api.UpdateTaskRequest, build.BuildConfig, bool) error
-	hydrateFromTask(context.Context, api.IAPIClient, *api.Task) error
+	fillInUpdateTaskRequest(*api.UpdateTaskRequest, build.BuildConfig, GetUpdateTaskRequestOptions) error
+	hydrateFromTask(api.Task, []api.ResourceMetadata) error
```
Comment (colinking) [inline pkg/deploy/taskdir/definitions/task_def.go:55]: Removed the pointer, as it wasn't necessary.

File: pkg/deploy/taskdir/definitions/task_def.go [hunk 2]
```diff
@@ -52,8 +51,8 @@ type Definition struct {
 }
 
 type taskKind interface {
-	fillInUpdateTaskRequest(context.Context, api.IAPIClient, *api.UpdateTaskRequest, build.BuildConfig, bool) error
-	hydrateFromTask(context.Context, api.IAPIClient, *api.Task) error
+	fillInUpdateTaskRequest(*api.UpdateTaskRequest, build.BuildConfig, GetUpdateTaskRequestOptions) error
```
Comment (colinking) [inline pkg/deploy/taskdir/definitions/task_def.go:54]: Switched to an options struct so it's more clear what the boolean is.

File: pkg/deploy/taskdir/definitions/task_def.go [hunk 3]
```diff
@@ -1276,7 +1235,16 @@ func (d Definition) taskKind() (taskKind, error) {
...
 }
 
-func (d Definition) GetUpdateTaskRequest(ctx context.Context, client api.IAPIClient, forBundle bool) (api.UpdateTaskRequest, error) {
+type GetUpdateTaskRequestOptions struct {
+	// List of resources that this task can attach.
+	AvailableResources []api.ResourceMetadata
+	// Set to `true` if this task is using bundle builds.
+	Bundle bool
+	// Set to `true` to silently ignore invalid definition fields.
+	IgnoreInvalid bool
```
Comment (francescaleung) [inline pkg/deploy/taskdir/definitions/task_def.go:1244]: What's the use case for setting this to true?

**Review by francescaleung — APPROVED** [review]

File: pkg/deploy/taskdir/definitions/task_def.go (hunk 3 above)
Comment (colinking) [reply pkg/deploy/taskdir/definitions/task_def.go:1244]: Definitions can be invalid when developing within studio. We use this method to generate a `Task` within the CLI.

**Comment by colinking** [comment]
@francescaleung based on https://github.com/airplanedev/cli/pull/982#discussion_r1123764042, I refactored `GetUpdateTaskRequest` to be `GetTask` and pulled down the `(Task) AsUpdateTaskRequest()` from an upstream PR. I'm going to go update the other PRs now.

**Comment by colinking** [comment]
@francescaleung my changes are in this commit specifically: https://github.com/airplanedev/lib/pull/568/commits/5af89a9c12082818ef3c37862cad29df2bca5a22

File: pkg/api/client.go [hunk 4]
```diff
@@ -56,6 +57,80 @@ type Task struct {
...
+
+	// Ensure all nullable fields are initialized since UpdateTaskRequest uses patch semantics.
+	if req.Kind == build.TaskKindImage {
+		if req.Command == nil {
+			req.Command = []string{}
+		}
+		if req.Arguments == nil {
+			req.Arguments = []string{}
+		}
+	}
```
Comment (francescaleung) [inline pkg/api/client.go:101]: this feels like it should be in the kind-specific `copyToTask`, not here

**Review by francescaleung — APPROVED** [review]

**Pull request merged**